### ISSUE TITLE
Allow to delete object props in extendObject

### ISSUE THREAD

**Issue by foxriver76** [ioBroker/ioBroker.js-controller#1735]
probably by setting null
see also #1730

**Comment by mcm1957** [comment]
At current telegram discussions the following problem arised which seems to be related to this issue.

If a state has parameter states set to {opt1:"opt1", opt2:"opt2"} its NOT possible to remove i.e. opt1 by using extendObject specifying states as {opt2:"opt2"}.

If extendObject has the paramater states specified its content should be written to the state object like a new value for i.e. units is merged into the existing state object

**Comment by foxriver76** [comment]
Yes, as we perform a kind of deep `extend` you could only overwrite properties, you can never delete. It would still be an option to delete via `null` of course you can then not set `null` in an `extend` call. Maybe we also find a smarter solution with e.g. an option, but could also be ugly because of the potential nesting.

**Comment by mcm1957** [comment]
OK, 
I'm not sure if deep extend is really needed here.

Personally I would like to see that paramaters ('first level attributes' like unit, name, description and states) are either left unchanged (if not present within extendObject object) or completly replaced by the data attached to them. I do not see any realistic situation where i.e. single language must be replaced or added and I think in 99,9% of usage the complete states data should be replaced.

And yes, an option cpould be a way to solve this. Or you could add updateObject which operates nearly like extendObject but replaces the parameters completly. (name suggested by others :-) )

**Comment by foxriver76** [comment]
The problem is at which level do you make the cut between merging and overriding. E.g. currently people set for example `common.name` for extending, so it is already no longer on top level. And in native, we have also have no idea, what's in there.

But for most things replacing stuff on second level (inside common, native) would be a good option.  Maybe we introduce a option `deepMerge` which is then `true` if not given, to stay backward compatible.

**Comment by mcm1957** [comment]
I would assume that nearly nobody uses extednObject with no common attribute or even without any common.*-
So I would replace level 2 (?) or lower as I stated above.

If common.name is present replace it (= current)
if common.unit is present replace it (= current)
if common.desc is present replace complete object (changed) or string (current)
if common.states is present replace complete object (changed)

Adding a deepMerge option to ensure compatibility would be a good idea. 

Is there any (legal) common.* construct which is deeper than 3 ? Mayby the option could accept a level value - but it might be too confusing as I do not really know how to count and what you refer to with level 1, 2, 3 ...

With native part this could get much more complex as real removel at level 1 could be desired too

Maybe a complete new function would be needed to as removing and recreating an object would be sometimes a good option but is not really possible, as this deletes all userOptions like history. So a sort of recreateObject completly as specified during setObject BUT keep id and userOptions would be helpful in some situations too.
